core/src/main/scala/io/gearpump/util/Graph.scala

```diff
@@ -254,6 +254,90 @@ class Graph[N, E](vertexList: List[N], edgeList: List[(N, E, N)]) extends Serial
   }
 
   /**
+   * Return all circles in graph.
+   */
+  def findCircles: mutable.Set[mutable.Set[N]] = {
+    val inStack = mutable.Map[N, Boolean]()
+    val stack = mutable.Stack[N]()
+    val DFN = mutable.Map[N, Int]()
+    val LOW = mutable.Map[N, Int]()
+    var index = 0
+
+    val circles = mutable.Set.empty[mutable.Set[N]]
+
+    def tarjan(node: N): Unit = {
+      DFN(node) = index
+      LOW(node) = index
+      index += 1
+      inStack(node) = true
+      stack.push(node)
+
+      outgoingEdgesOf(node).foreach {
+        edge => {
+          if (!DFN.contains(edge._3)) {
+            tarjan(edge._3)
+            if (LOW.get(edge._3).get < LOW.get(node).get)
+              LOW(node) = LOW(edge._3)
+          } else {
+            if (inStack.get(edge._3).get && (DFN.get(edge._3).get < LOW.get(node).get))
+              LOW(node) = DFN(edge._3)
+          }
+        }
+      }
+
+      if (DFN.get(node).get == LOW.get(node).get) {
+        val set = mutable.Set[N]()
+        var n: N = node
+        do {
+          n = stack.pop()
+          inStack(n) = false
+          set.add(n)
+        } while (n != node)
+        circles.add(set)
+      }
+    }
+
+    _vertices.foreach {
+      node => {
+        if (!DFN.contains(node)) tarjan(node)
+      }
+    }
+
+    circles
+  }
+
+  /**
+   * Return an iterator of vertex in topological order of graph with circles
+   * The node returned by Iterator is stable sorted.
+   */
+  def topologicalOrderWithCirclesIterator: Iterator[N] = {
```

Comment:
Are you sure it is "stable sorted"?

Since you are using Set, and Set is not stable when expanded.